Revision of the advanced topics document. (#354)
epinio/docs: docs/explanations/advanced.md

---
sidebar_label: "Advanced topics"
title: "Advanced topics"
---

## Prerequisites

There are components (Ingress controller and a Certificate Manager) required in a Kubernetes cluster before
an Epinio installation.
The Epinio helm chart doesn't deploy these components.

### Ingress controller

On a Kubernetes cluster, certain services need to be reachable from outside the cluster.
For Epinio, the `API server` is one of them.
Certain of the applications deployed with Epinio might also need to be reachable from outside.

One way to expose services externally is by creating [Ingress resources](https://kubernetes.io/docs/concepts/services-networking/ingress/).
Ingress resources on their own, have no effect.

> They're just descriptions of what the needed routing.

Reviewer: ```suggestion
They're just descriptions of the needed routing.
```

Author: Thanks.

An Ingress controller handles the implementation of this routing.

:::note

Most clusters have an Ingress controller deployed by default.
Install [Traefik](https://doc.traefik.io/traefik/providers/kubernetes-ingress/), or another, if an Ingress controller is absent.
Traefik is used in Epinio CI to test Epinio.

There is installation documentation for [Traefik](../installation/install_epinio.md#ingress-controller)

:::

Epinio creates an Ingress resource for the Epinio API server and each application deployed.

### Cert-manager

[Cert-manager documentation](https://cert-manager.io/docs/)

Cert-manager is a Kubernetes controller that generates and renews certificates.
You need these certificates  to securely serve the endpoints over TLS (for example, the Epinio API server).

> Epinio supports options when it comes to certificate issuers (Let's Encrypt, private your own CA, self signed certs).

Reviewer: ```suggestion
Epinio supports options when it comes to certificate issuers (Let's Encrypt, your own private CA, self signed certs).
```

Author: Thanks.

### Epinio API server

The main component of Epinio is the API server.
The same `epinio` binary is both the server and the CLI.
The server starts with the `epinio server` command within a Kubernetes Pod, configured for Epinio.

Epinio CLI and web UI functionality are implemented using the endpoints provided by the Epinio API server component.
For example, when the user asks Epinio to "push" an application,
the CLI contacts the "Upload", "Stage" and "Deploy" endpoints of the Epinio API to:
- upload the application code
- create a container image for the application using this code
- run the application on the cluster.

The Epinio API server runs on the cluster and is reached using Kubernetes resources like
Deployments, Services,  Ingresses and Secrets.

### Kubed

[Kubed documentation](https://github.com/kubeops/kubed)

Epinio installs Kubed to keep secrets, needed in more than one namespace, synced.
For example, the image pull secret is needed in every application namespace
so that Kubernetes can pull the built application images from the [Container Registry](#container-registry).

Kubed makes sure that if the source secret changes, the copies change too.

> <!--TODO: Still the case?-->

Reviewer: AFAIK yes. @enrichman ?

Author: Hi @enrichman , Just to be clear i@m asking about whether it's still the case that 'Secret rotation' will be handled by Epinio in the future. Thanks,

Reviewer 2: I'm not sure to understand which is the secrets we are referring to, and what kind of rotation. Regarding Kubed and the registry that's probably already the case. For other secrets I don't think that we are going to manage rotations somehow.

You can see a [How-to here](../howtos/customization/setup_external_s3.md).

### s3gw

[The s3gw project link](https://github.com/aquarist-labs/s3gw)

S3gw is a lightweight S3-compatible solution.
You can specify it in the Epinio Helm chart as an alternative to Minio or an external S3 provider.

### Container Registry

The result of Epinio's application staging is a container image.
You use this image to create a Kubernetes deployment to run the application code.
The staging job writes the image to a container registry.
There is further information in the [detailed push process](detailed-push-process.md) documentation.

By default, the Epinio installation deploys a container registry inside the Kubernetes cluster, making the process fast and convenient.

Epinio comes with two consumers of this registry:

1. Staging job - pushing the images
2. Kubernetes - pulling the images when creating a deployment for the application

<!--TODO:Did I reword this correctly. I think it needs some better explanation. Who can help here?-->
All consumers should communicate with the registry using TLS to encrypt communication.

> Epinio controls the staging job and it ensures that iit trusts the CA used to sign the registry certificate.

Reviewer: ```suggestion
Epinio controls the staging job and it ensures that it trusts the CA used to sign the registry certificate.
```

> Achieving the same for Kubernetes requires configuration that is impossible in the cluster,

Reviewer: ```suggestion
Achieving the same for Kubernetes requires configuration that is impossible from within the cluster,
```
```suggestion
Achieving the same for Kubernetes requires configuration that is impossible from inside the cluster,
```
I.e. an application like E, running in a pod of the cluster does not and cannot have the permissions to change things so highly relevant to cluster security. Such changes can only be made from the outside, by a cluster operator.

Author: Thanks.

therefore Epinio has no way to ensure that it trusts the CA.

There are 3 options:

1. Let the Epinio user manually configure Kubernetes to trust the CA
1. Use a well-known trusted CA, so there's no configuration needed
1. Don't encrypt the communication at all

> <!--TODO:Doesn't sound that great to me. What's happening moving forward, is the current situation accurately described.-->

Reviewer: AFAICT the situation is described correctly. And, AFAIK, we have no way forward out of this.

Author: Thanks.
